Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidator.java

```diff
@@ -0,0 +1,213 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE UNITED STATES GOVERNMENT BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package gov.hhs.fha.nhinc.callback.cxf;
+
+import gov.hhs.fha.nhinc.callback.openSAML.CertificateManager;
+import gov.hhs.fha.nhinc.callback.openSAML.CertificateManagerImpl;
+import java.security.InvalidKeyException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.PublicKey;
+import java.security.SignatureException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.commons.lang.StringUtils;
+import org.apache.ws.security.WSSecurityException;
+import org.apache.ws.security.components.crypto.DERDecoder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Validate chain of trust
+ *
+ * @author mpnguyen
+ *
+ */
+public class CertificateChainValidator {
+    private static final Logger LOG = LoggerFactory.getLogger(CertificateChainValidator.class);
+    private final CertificateManager certificateManager;
+
+    // http://docs.oracle.com/javase/7/docs/api/java/security/cert/X509Extension.html#getExtensionValue(java.lang.String)
+    private static final String AUTHORITY_KEY_ID = "2.5.29.35";
+    private static final String SUBJECT_KEY_ID = "2.5.29.14";
+
+    /**
+     *
+     */
+    private CertificateChainValidator() {
+        certificateManager = CertificateManagerImpl.getInstance();
+    }
+
+    /**
+     * @param certificateManager
+     */
+    CertificateChainValidator(CertificateManager certificateManager) {
+        this.certificateManager = certificateManager;
+    }
+    public static CertificateChainValidator getInstance() {
+        return new CertificateChainValidator();
+    }
+
+    /**
+     * ALL root CA are self-sign cert. Verify if cert is self-sign or chain of trust
+     *
+     * @param certToCheck
+     * @return true if self-sign cert or false if chain of trust
+     */
+    public boolean isPartChainOfTrust(X509Certificate certToCheck) {
+        try {
+            // Try to verify certificate signature with its own public key
+            PublicKey key = certToCheck.getPublicKey();
+            certToCheck.verify(key);
+            return false;
```

Comment:
Maybe method name is confused.  Only self-sign cert can verify itself with its public key.  This operation allows us to determine if cert is self-sigh cert (or root) or sign by something else.